### DISCUSSION TITLE
Just training NER

### DISCUSSION THREAD

**Discussion by dakinggg** [explosion/spaCy#6649] [Help: Coding & Implementations] [answered]
(This might be a duplicate of #6548, but there were a couple questions on that thread)
I have a model with parser and tagger (and tok2vec) trained, and I would like to train NER on top of that. So in my config I have
```
[components]

[components.ner]
factory = "ner"
moves = null
update_with_oracle_cut_size = 100

[components.ner.model]
@architectures = "spacy.TransitionBasedParser.v1"
state_type = "ner"
extra_state_tokens = false
hidden_width = 64
maxout_pieces = 2
use_upper = true
nO = null

[components.ner.model.tok2vec]
@architectures = "spacy.Tok2VecListener.v1"
# TODO: How to get this programatically when tok2vec is loaded from source?
width = 256
upstream = "*"

[components.parser]
source = "output/en_core_sci_sm_parser_tagger/model-best"

[components.tagger]
source = "output/en_core_sci_sm_parser_tagger/model-best"

[components.tok2vec]
source = "output/en_core_sci_sm_parser_tagger/model-best"

...

[training]
dev_corpus = "corpora.dev"
train_corpus = "corpora.train"
seed = ${system.seed}
gpu_allocator = ${system.gpu_allocator}
dropout = 0.1
accumulate_gradient = 1
patience = 1600
max_epochs = 7
# TODO: set back to 0 when doing for real
max_steps = 400
eval_frequency = 200
frozen_components = ["parser", "tagger"]
before_to_disk = null
```

and then I run  `spacy train configs/ner_sm.cfg --output output/en_core_sci_sm_ner --code scispacy/base_project_code.py`. This results in 
```
=========================== Initializing pipeline ===========================
Set up nlp object from config
Pipeline: ['tok2vec', 'tagger', 'parser', 'ner']
Resuming training for: ['tok2vec']
Created vocabulary
Finished initializing nlp object
Initialized pipeline components: ['ner']
✔ Initialized pipeline

============================= Training pipeline =============================
ℹ Pipeline: ['tok2vec', 'tagger', 'parser', 'ner']
ℹ Frozen components: ['parser', 'tagger']
```
which looks right, but then errors out with `ValueError: [E954] The Tok2Vec listener did not receive any valid input from an upstream component.`. Not sure if this is a bug, or if I am missing something, but appreciate the help! Also, is the vocab all handled when loading a component using `source`? Or do I need to initialize the vocab to the vocab output from parser/tagger as well (and how do i do that)? Thanks!

**Comment by svlandeg** [comment]
Hm, this feels like a bug to me. Generally, the `Tok2VecListener` obtains its token vectors from an upstream `tok2vec` component, that should be run before it in the pipeline. If that `tok2vec` indeed runs (which it seems like), then I wonder why the `Tok2VecListener` can't access its results.

**Comment by svlandeg** [comment]
By the way - are your `parser` and `tagger` trained on the same `tok2vec`? Because if you freeze those components, but not the `tok2vec` they rely on, then the final results after all training is done, may be quite bad for the frozen parser and tagger, I would think.

**Comment by dakinggg** [comment, marked as answer]
ah, good point. Is there a suggested way to train parser/tagger on one set of data, and NER on another set of data? Give NER its own tok2vec rather than sharing the same one as the parser and tagger perhaps?

**Reply by svlandeg** [reply]
Yes! With the new v3 config system, you should be able to have multiple `Tok2Vec` components in one pipeline. This will probably end up fixing your original problem (though we might still need to look at that in detail as the error message is still weird).

Basically, you can give your different tok2vec components different names, and then refer to the correct one in the `Tok2VecListener` with the `upstream` argument, cf here: https://nightly.spacy.io/api/architectures#Tok2VecListener. Then you can train tagger + parser + tok2vec_1 in a run, store them, and then train ner + tok2vec_2 in a second run or something like that.

Or, instead of having two listeners, you could also just have a separate Tok2Vec component that's only used by your NER. You can basically mix and match the strategies described here: https://nightly.spacy.io/usage/embeddings-transformers#embedding-layers

**Reply by dakinggg** [reply]
Thanks! I was able to just give the NER component its own tok2vec like so
```
[components.ner.model.tok2vec]
@architectures = "spacy.Tok2Vec.v1"

[components.ner.model.tok2vec.embed]
@architectures = "spacy.MultiHashEmbed.v1"
width = 256
attrs = ["NORM","PREFIX","SUFFIX","SHAPE"]
rows = [5000,2500,2500,2500]
include_static_vectors = true

[components.ner.model.tok2vec.encode]
@architectures = "spacy.MaxoutWindowEncoder.v1"
width = 256
depth = 8
window_size = 1
maxout_pieces = 3
```

It at least starts training without any problems. Is that the correct way to set it? So, as you said, I think I am unblocked, but the original error is likely still an issue (I'll let you decide whether to accept one of your answers, open an issue, etc. I don't really understand the protocols for "discussions"). Thanks again!

**Reply by honnibal** [reply]
Yes it makes sense to have the NER model "own" its `tok2vec` in this sort of situation. So what you have there looks good to me. The only extra piece is to make sure you "freeze" the other components in the pipeline so that they're not also updated.

**Reply by svlandeg** [reply]
>I don't really understand the protocols for "discussions"

No worries, we're still figuring that out ourselves ;-)

**Comment by mehmetilker** [comment]
Hi @danielkingai2,

Did you get training result?
Any problem with tagger/parser and NER accuracy?

**Reply by dakinggg** [reply]
I haven't tested thoroughly, but it seemed fine. NER with its own tok2vec shouldn't affect the parser/tagger at all, and for NER it should be similar to training them all separately as in previous spacy versions.

**Reply by myeghaneh** [reply]
hi, can you test it? mine it works, but the NER score drop down, however it seems the parser got better!

I explained here

https://github.com/explosion/spaCy/discussions/6714#discussioncomment-280228
that is a bit weird, as we only used the pre-trained model for parser it should not affect the NER 

Do you have any idea why is so?

**Reply by mehmetilker** [reply]
In my case I see lower tagger/parser score. Even though I have them in frozen_components.

**Reply by myeghaneh** [reply]
thank you for your message, I have also seen the same,  however, I use the same train data, as soon as I add a parser to result got lower, 
 :(  around 3 percent which makes a huge change in my pipeline, 
 
 I tried different approaches, add a parser to ''disable" then I do NOT know how to enable them
 
  do you have any innovative idea to make it work, can not we train NER and then add parser some how externally?

**Reply by mehmetilker** [reply]
Take a look at my config: https://github.com/explosion/spaCy/discussions/6548#discussioncomment-295218

I think problem is configuration related with sharing of tov2vec (or transformer in my case)
I have tried different configuration like freezing transformer and provide previous transformer component as source but either it ended with configuration error or I got low score...

**Reply by dakinggg** [reply]
Hi, I have not had any problem with parser/tagger degrading after training NER, using my setup.

**Reply by mehmetilker** [reply]
Thanks for let as know. I could not succeed with transformer instead of tok2vec....

**Reply by myeghaneh** [reply]
as soon as I add tagger and parser as "frozen components" the result of NER
drop down a bit (F1 .94)  in comparison with a model with pure NER (F1 .098)
however, the problem is not only this, but the Parser is not
also good as my ore-trained model en-core-web-sm

I am trying a different set ups for config file

**Reply by myeghaneh** [reply]
@danielkingai2

could you share your config file (filled and  base config file) as I guess we are working on the same issue

I am training a NER on top of pre-exciting Parser and Tagger of en-core-web-sm model, which mean that I have training data for my custom NER. and trained my model with frozen parser and tagger 

but there is 2 problems 
1- the result of parser and tagger is not as en-core-web-sm model,
2- the result of NER drop down 4 percent after adding tagger and parser as frozen component 
many thanks

**Reply by dakinggg** [reply]
https://github.com/allenai/scispacy/blob/aaa349c1dda784ef7d6877e277619c13bf6fd77a/configs/base_ner.cfg
